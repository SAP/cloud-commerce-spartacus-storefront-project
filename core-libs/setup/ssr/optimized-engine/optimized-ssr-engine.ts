/* webpackIgnore: true */
import { Request, Response } from 'express';
import * as fs from 'fs';
import { NgExpressEngineInstance } from '../engine-decorator/ng-express-engine-decorator';
import { RenderingCache } from './rendering-cache';
import {
  RenderingStrategy,
  SsrOptimizationOptions,
} from './ssr-optimization-options';

/**
 * The rendered pages are kept in memory to be served on next request. If the `cache` is set to `false`, the
 * response is evicted as soon as the first successful response is successfully returned.
 */
export class OptimizedSsrEngine {
  protected currentConcurrency = 0;
  protected renderingCache = new RenderingCache(this.ssrOptions);
  private templateCache = new Map<string, string>();

  get engineInstance(): NgExpressEngineInstance {
    return this.renderResponse.bind(this);
  }

  constructor(
    protected expressEngine: NgExpressEngineInstance,
    protected ssrOptions?: SsrOptimizationOptions
  ) {}

  /**
   * When SSR page can not be returned in time, we're returning index.html of
   * the CSR application.
   * The CSR application is returned with the "Cache-Control: no-store" response-header. This notifies external cache systems to not use the CSR application for the subsequent request.
   */
  protected fallbackToCsr(
    response: Response,
    filePath: string,
    callback: (err?: Error | null, html?: string) => void
  ) {
    response.set('Cache-Control', 'no-store');
    callback(undefined, this.getDocument(filePath));
  }

  protected getRenderingKey(request: Request): string {
    return this.ssrOptions?.renderKeyResolver
      ? this.ssrOptions.renderKeyResolver(request)
      : request.originalUrl;
  }

  protected getRenderingStrategy(request: Request): RenderingStrategy {
    return this.ssrOptions?.renderingStrategyResolver
      ? this.ssrOptions.renderingStrategyResolver(request)
      : RenderingStrategy.DEFAULT;
  }

  protected shouldRender(request: Request): boolean {
    const concurrencyLimitExceed = this.ssrOptions?.concurrency
      ? this.currentConcurrency >= this.ssrOptions.concurrency
      : false;

    const isRendering = this.renderingCache.isRendering(
      this.getRenderingKey(request)
    );

    if (isRendering) {
      this.log(`CSR fallback: rendering in progress (${request?.originalUrl})`);
    } else if (concurrencyLimitExceed) {
      this.log(
        `CSR fallback: Concurrency limit exceeded (${this.ssrOptions.concurrency})`
      );
    }

    return (
      (!isRendering &&
        !concurrencyLimitExceed &&
        this.getRenderingStrategy(request) !== RenderingStrategy.ALWAYS_CSR) ||
      this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
    );
  }

  protected shouldTimeout(request: Request): boolean {
    return (
      !!this.ssrOptions?.timeout ||
      this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
    );
  }

  protected getTimeout(request: Request): number {
    return this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
      ? this.ssrOptions?.forcedSsrTimeout ?? 60000
      : this.ssrOptions?.timeout ?? 0;
  }

  protected returnCachedRender(
    request: Request,
    callback: (err?: Error | null, html?: string) => void
  ): boolean {
    const key = this.getRenderingKey(request);

    if (this.renderingCache.isReady(key)) {
      const cached = this.renderingCache.get(key);
      callback(cached.err, cached.html);

      if (!this.ssrOptions?.cache) {
        // we drop cached rendering if caching is disabled
        this.renderingCache.clear(key);
      }
      return true;
    }
    return false;
  }

  protected renderResponse(
    filePath: string,
    options: any,
    callback: (err?: Error | null, html?: string) => void
  ): void {
    const request: Request = options.req;
    const response: Response = options.res || options.req.res;

    const renderingKey = this.getRenderingKey(request);

    if (!this.returnCachedRender(request, callback)) {
      if (this.shouldRender(request)) {
        this.currentConcurrency++;
        let waitingForRender;

        if (this.shouldTimeout(request)) {
          // establish timeout for rendering
          const timeout = this.getTimeout(request);
          waitingForRender = setTimeout(() => {
            waitingForRender = undefined;
            this.fallbackToCsr(response, filePath, callback);
            this.log(
              `SSR rendering exceeded timeout ${timeout}, fallbacking to CSR for ${request?.originalUrl}`,
              false
            );
          }, timeout);
        } else {
          this.fallbackToCsr(response, filePath, callback);
        }

        // start rendering
        this.renderingCache.setAsRendering(renderingKey);

        // setting the timeout for hanging renders that might not ever finish due to various reasons
        // releasing concurrency slots by decreasing the `this.currentConcurrency--`.
        let maxRenderTimeout: NodeJS.Timeout | undefined = setTimeout(() => {
          this.currentConcurrency--;
          this.renderingCache.clear(renderingKey);
          maxRenderTimeout = undefined;

          this.log(
            `Rendering of ${request?.originalUrl} was not able to complete. This might cause memory leaks!`,
            false
          );
        }, this.ssrOptions?.maxRenderTime ?? 300000); // 300000ms == 5 minutes

        this.log(`Rendering started (${request?.originalUrl})`);
        this.expressEngine(filePath, options, (err, html) => {
          if (!maxRenderTimeout) {
            // ignore this render's result because it exceeded maxRenderTimeout
            this.log(
              `Rendering of ${request.originalUrl} completed after the specified maxRenderTime, therefore it was ignored.`
            );
            return;
          }
          clearTimeout(maxRenderTimeout);
          this.currentConcurrency--;

          this.log(`Rendering completed (${request?.originalUrl})`);

          if (waitingForRender) {
            // if request is still waiting for render, return it
            clearTimeout(waitingForRender);
            callback(err, html);

            // store the render only if caching is enabled
            if (this.ssrOptions?.cache) {
              this.renderingCache.store(renderingKey, err, html);
            } else {
              this.renderingCache.clear(renderingKey);
            }
          } else {
            // store the render for future use
            this.renderingCache.store(renderingKey, err, html);
          }
        });
      } else {
        // if there is already rendering in progress, return the fallback
        this.fallbackToCsr(response, filePath, callback);
      }
    } else {
      this.log(`Render from cache (${request?.originalUrl})`);
    }
  }

  protected log(message: string, debug = true) {
    if (!debug || this.ssrOptions.debug) {
      console.log(message);
    }
  }

  /** Retrieve the document from the cache or the filesystem */
  protected getDocument(filePath: string): string {
    let doc = this.templateCache.get(filePath);

    if (!doc) {
      // fs.readFileSync could be missing in a browser, specifically
      // in a unit tests with { node: { fs: 'empty' } } webpack configuration
      doc = fs?.readFileSync ? fs.readFileSync(filePath, 'utf-8') : '';
      this.templateCache.set(filePath, doc);
    }

    return doc;
  }
}
