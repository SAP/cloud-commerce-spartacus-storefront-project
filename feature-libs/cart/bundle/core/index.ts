export * from './connectors/index';
export * from './model/index';
export * from './bundle-core.module';
export * from './connectors/index';
export * from './facade/index';
export * from './store/actions/index';
export * from './store/selectors/index';
export * from './store/bundle-state';
export * from './services/index';
