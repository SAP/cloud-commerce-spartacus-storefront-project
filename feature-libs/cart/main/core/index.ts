export { CartCoreModule } from './cart-core.module';
export { CartPersistenceModule } from './cart-persistence.module';
export * from './connectors/index';
export * from './event/index';
export * from './facade/index';
export * from './services/index';
export * from './store/actions/index';
export * from './store/multi-cart-state';
export * from './store/selectors/index';
export * from './utils/utils';
