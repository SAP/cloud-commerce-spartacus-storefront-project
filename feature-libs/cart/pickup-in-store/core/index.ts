export * from './facades';
export * from './pickup-in-store-core.module';
