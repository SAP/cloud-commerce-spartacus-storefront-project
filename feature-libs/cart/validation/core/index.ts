export * from './cart-validation-core.module';
export * from './connectors/index';
export * from './facade/index';
export * from './model/index';
