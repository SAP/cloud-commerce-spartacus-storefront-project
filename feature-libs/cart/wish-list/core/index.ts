export * from './facade/index';
export * from './store/actions/index';
export * from './store/selectors/index';
export * from './store/wish-list-state';
export * from './utils/utils';
export * from './wish-list-core.module';
