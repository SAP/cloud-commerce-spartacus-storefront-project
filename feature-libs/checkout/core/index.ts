export * from './checkout-core.module';
