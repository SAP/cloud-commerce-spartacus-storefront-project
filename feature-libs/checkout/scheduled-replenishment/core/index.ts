export * from './checkout-scheduled-replenishment-core.module';
export * from './connectors/index';
export * from './facade/index';
