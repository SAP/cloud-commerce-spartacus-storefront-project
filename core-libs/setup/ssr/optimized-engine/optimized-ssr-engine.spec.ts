import { fakeAsync, flush, tick } from '@angular/core/testing';
import { OptimizedSsrEngine } from './optimized-ssr-engine';
import {
  RenderingStrategy,
  SsrOptimizationOptions,
} from './ssr-optimization-options';

/**
 * Helper class to easily create and test engine wrapper against mocked engine.
 *
 * Mocked engine will return sample rendering after 100 milliseconds.
 *
 * Usage:
 * 1. Instantiate the class with desired options
 * 2. Call request() to run request through engine
 * 3. Examine renders property for the renders
 */
class TestEngineRunner {
  /** Accumulates html output for engine runs */
  renders: string[] = [];

  /** Accumulates response parameters for engine runs */
  responseParams: object[] = [];

  renderCount = 0;
  optimizedSsrEngine: OptimizedSsrEngine;
  engineInstance;

  constructor(options: SsrOptimizationOptions, renderTime?: number) {
    // mocked engine instance that will render test output in 100 milliseconds
    const engineInstanceMock = (
      filePath: string,
      _: any,
      callback: Function
    ) => {
      setTimeout(() => {
        callback(undefined, `${filePath}-${this.renderCount++}`);
      }, renderTime ?? 100);
    };

    this.optimizedSsrEngine = new OptimizedSsrEngine(
      engineInstanceMock,
      options
    );
    this.engineInstance = this.optimizedSsrEngine.engineInstance;
  }

  /** Run request against the engine. The result will be stored in rendering property. */
  request(url: string) {
    const response = {};
    const optionsMock = {
      req: {
        originalUrl: url,
      },
      res: {
        set: (key: string, value: any) => (response[key] = value),
      },
    };

    this.engineInstance(url, optionsMock, (_, html) => {
      this.renders.push(html);
      this.responseParams.push(response);
    });

    return this;
  }
}

describe('OptimizedSsrEngine', () => {
  describe('timeout option', () => {
    it('should fallback to csr if rendering exceeds timeout', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({ timeout: 50 }).request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['']);
    }));

    it('should return timed out render in the followup request', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({ timeout: 50 }).request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['']);

      engineRunner.request('a');
      expect(engineRunner.renders[1]).toEqual('a-0');
    }));

    it('should return render if rendering meets timeout', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({ timeout: 150 }).request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['a-0']);
    }));

    it('should fallback instantly if is set to 0', () => {
      const engineRunner = new TestEngineRunner({ timeout: 0 }).request('a');
      expect(engineRunner.renders).toEqual(['']);
    });
  });

  describe('no-store cache controll header', () => {
    it('should be applied for a fallback', () => {
      const engineRunner = new TestEngineRunner({ timeout: 0 }).request('a');
      expect(engineRunner.renders).toEqual(['']);
      expect(engineRunner.responseParams).toEqual([
        { 'Cache-Control': 'no-store' },
      ]);
    });

    it('should not be applied for a render within time limit', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({ timeout: 200 }).request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['a-0']);
      expect(engineRunner.responseParams).toEqual([{}]);
    }));

    it('should not be applied for a render served with next response', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({ timeout: 50 }).request('a');
      tick(200);
      engineRunner.request('a');
      expect(engineRunner.renders).toEqual(['', 'a-0']);
      expect(engineRunner.responseParams).toEqual([
        { 'Cache-Control': 'no-store' },
        {},
      ]);
    }));
  });

  describe('cache option', () => {
    it('should not cache requests if disabled', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        cache: false,
        timeout: 200,
      }).request('a');

      tick(200);
      engineRunner.request('a');
      tick(200);
      engineRunner.request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['a-0', 'a-1', 'a-2']);
    }));

    it('should cache requests if enabled', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        cache: true,
        timeout: 200,
      }).request('a');

      tick(200);
      engineRunner.request('a');
      tick(200);
      engineRunner.request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['a-0', 'a-0', 'a-0']);
    }));
  });

  describe('concurrency option', () => {
    it('should limit concurrency and fallback to csr', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        concurrency: 3,
        timeout: 200,
      })
        .request('a')
        .request('b')
        .request('c')
        .request('d')
        .request('e');

      tick(200);
      expect(engineRunner.renders).toEqual(['', '', 'a-0', 'b-1', 'c-2']);
    }));

    it('should reinvigorate limit after emptying the queue', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        concurrency: 2,
        timeout: 200,
      }).request('a');

      tick(60);
      engineRunner.request('b').request('c');
      tick(60);
      engineRunner.request('d').request('e');
      tick(200);
      engineRunner.request('f').request('g');
      tick(200);

      expect(engineRunner.renders).toEqual([
        '',
        'a-0',
        '',
        'b-1',
        'd-2',
        'f-3',
        'g-4',
      ]);
    }));
  });

  describe('ttl option', () => {
    it('should invalidate expired renders', fakeAsync(() => {
      let currentDate = 100;
      spyOn(Date, 'now').and.callFake(() => currentDate);

      const engineRunner = new TestEngineRunner({
        cache: true,
        ttl: 300,
        timeout: 200,
      }).request('a');

      tick(200);
      currentDate += 200;
      engineRunner.request('a');

      tick(200);
      currentDate += 200;
      engineRunner.request('a');

      tick(200);
      expect(engineRunner.renders).toEqual(['a-0', 'a-0', 'a-1']);
    }));
  });

  describe('renderKeyResolver option', () => {
    it('should use custom render key resolver', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        renderKeyResolver: (req) => req.originalUrl.substr(0, 2),
        timeout: 200,
        cache: true,
      }).request('ala');

      tick(200);
      engineRunner.request('ale');
      tick(200);
      engineRunner.request('ela');
      tick(200);
      engineRunner.request('alu');
      tick(200);
      engineRunner.request('elu');
      tick(200);
      expect(engineRunner.renders).toEqual([
        'ala-0',
        'ala-0',
        'ela-1',
        'ala-0',
        'ela-1',
      ]);
    }));
  });

  describe('renderingStrategyResolver option', () => {
    it('always SSR should ignore timeout', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        renderingStrategyResolver: () => RenderingStrategy.ALWAYS_SSR,
        timeout: 50,
        cache: true,
      }).request('a');

      tick(200);
      expect(engineRunner.renders).toEqual(['a-0']);
    }));

    it('always CSR should return CSR instantly', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        renderingStrategyResolver: () => RenderingStrategy.ALWAYS_CSR,
        timeout: 200,
        cache: true,
      }).request('a');

      tick(200);
      engineRunner.request('a');
      tick(200);
      expect(engineRunner.renders).toEqual(['', '']);
    }));

    it('default should obey the timeout', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        renderingStrategyResolver: () => RenderingStrategy.DEFAULT,
        timeout: 50,
      }).request('a');

      tick(200);
      engineRunner.request('a');
      expect(engineRunner.renders).toEqual(['', 'a-0']);
    }));
  });

  describe('forcedSsrTimeout option', () => {
    it('should fallback to csr for always ssr rendering strategy', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        renderingStrategyResolver: () => RenderingStrategy.ALWAYS_SSR,
        timeout: 50,
        forcedSsrTimeout: 80,
      }).request('a');

      tick(60);
      expect(engineRunner.renders).toEqual([]);

      tick(50);
      expect(engineRunner.renders).toEqual(['']);

      engineRunner.request('a');
      expect(engineRunner.renders).toEqual(['', 'a-0']);
    }));

    it('should not affect default rendering strategy', fakeAsync(() => {
      const engineRunner = new TestEngineRunner({
        timeout: 50,
        forcedSsrTimeout: 80,
      }).request('a');

      tick(60);
      expect(engineRunner.renders).toEqual(['']);

      tick(50);
      engineRunner.request('a');
      expect(engineRunner.renders).toEqual(['', 'a-0']);
    }));
  });

  describe('maxRenderTime option', () => {
    const fiveMinutes = 300000;

    it('should not kick-in for the non-hanging (normal) renders', fakeAsync(() => {
      const renderTime = 10;
      const requestUrl = 'a';
      const engineRunner = new TestEngineRunner({}, renderTime).request(
        requestUrl
      );
      spyOn<any>(engineRunner.optimizedSsrEngine, 'log').and.callThrough();

      tick(renderTime + 1);
      expect(engineRunner.renderCount).toEqual(1);
      expect(engineRunner.optimizedSsrEngine['log']).not.toHaveBeenCalledWith(
        `Rendering of ${requestUrl} was not able to complete. This might cause memory leaks!`,
        false
      );
    }));

    it('should use the default value of 5 minutes for hanging renders', fakeAsync(() => {
      const requestUrl = 'a';
      const renderTime = fiveMinutes + 100;
      const engineRunner = new TestEngineRunner({}, renderTime).request(
        requestUrl
      );
      spyOn<any>(engineRunner.optimizedSsrEngine, 'log').and.callThrough();

      tick(fiveMinutes);
      expect(engineRunner.renderCount).toEqual(0);
      expect(engineRunner.optimizedSsrEngine['log']).toHaveBeenCalledWith(
        `Rendering of ${requestUrl} was not able to complete. This might cause memory leaks!`,
        false
      );

      tick(101);
      expect(engineRunner.renderCount).toEqual(1);
    }));

    it('should use the provided value instead of the default one', fakeAsync(() => {
      const requestUrl = 'a';
      const renderTime = 200;
      const maxRenderTime = renderTime - 50; // shorter than the predicted render time
      const engineRunner = new TestEngineRunner(
        { maxRenderTime },
        renderTime
      ).request(requestUrl);
      spyOn<any>(engineRunner.optimizedSsrEngine, 'log').and.callThrough();

      tick(maxRenderTime);
      expect(engineRunner.renderCount).toEqual(0);
      expect(engineRunner.optimizedSsrEngine['log']).toHaveBeenCalledWith(
        `Rendering of ${requestUrl} was not able to complete. This might cause memory leaks!`,
        false
      );

      tick(50);
      expect(engineRunner.renderCount).toEqual(1);
    }));

    it('should release the concurrency slot for the hanging render', fakeAsync(() => {
      const hangingRequest = 'a';
      const csrRequest = 'b';
      const ssrRequest = 'c';
      const renderTime = 200;
      const maxRenderTime = renderTime - 50; // shorter than the predicted render time
      const engineRunner = new TestEngineRunner(
        { concurrency: 1, maxRenderTime },
        renderTime
      );
      spyOn<any>(engineRunner.optimizedSsrEngine, 'log').and.callThrough();

      // issue two requests
      engineRunner.request(hangingRequest);
      engineRunner.request(csrRequest);
      expect(engineRunner.optimizedSsrEngine['currentConcurrency']).toEqual(1);

      tick(1);
      // while the concurrency slot is busy rendering the first hanging request, the second request gets the CSR version
      expect(engineRunner.optimizedSsrEngine['log']).toHaveBeenCalledWith(
        `CSR fallback: Concurrency limit exceeded (1)`
      );
      expect(engineRunner.renderCount).toEqual(0);
      expect(engineRunner.optimizedSsrEngine['currentConcurrency']).toEqual(1);

      tick(maxRenderTime);
      expect(engineRunner.optimizedSsrEngine['log']).toHaveBeenCalledWith(
        `Rendering of ${hangingRequest} was not able to complete. This might cause memory leaks!`,
        false
      );
      expect(engineRunner.renderCount).toEqual(0);

      // even though the hanging request is still rendering, we've freed up a slot for a new request
      engineRunner.request(ssrRequest);
      tick(1);
      expect(engineRunner.optimizedSsrEngine['log']).toHaveBeenCalledWith(
        `Rendering started (${ssrRequest})`
      );
      expect(engineRunner.optimizedSsrEngine['currentConcurrency']).toEqual(1);

      flush();
    }));
  });
});
