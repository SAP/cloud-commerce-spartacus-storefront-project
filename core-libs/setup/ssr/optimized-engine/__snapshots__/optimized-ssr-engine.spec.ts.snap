// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptimizedSsrEngine avoidCachingErrors option when using default cacheStrategyResolver should cache errors if \`avoidCachingErrors\` is set to false 1`] = `
[
  [Error: a-0],
  [Error: a-0],
  [Error: a-0],
]
`;

exports[`OptimizedSsrEngine avoidCachingErrors option when using default cacheStrategyResolver should not cache errors if \`avoidCachingErrors\` is set to true 1`] = `
[
  [Error: a-0],
  [Error: a-1],
  [Error: a-2],
]
`;

exports[`OptimizedSsrEngine cacheStrategyResolver option should cache errors if \`cacheStrategyResolver\` returns true 1`] = `
[
  [Error: a-0],
  [Error: a-0],
  [Error: a-0],
]
`;

exports[`OptimizedSsrEngine cacheStrategyResolver option should not cache errors if \`cacheStrategyResolver\` returns false 1`] = `
[
  [Error: a-0],
  [Error: a-1],
  [Error: a-2],
]
`;
